File: tron_unity/Assets/Scripts/Heuristic.cs
using System.Collections.Generic;
using UnityEngine;

public static class Heuristic
{
    public static readonly Vector2Int[] DIRS = { Vector2Int.up, Vector2Int.right, Vector2Int.down, Vector2Int.left };

    public static float ChamberHeuristic(int[,] walls, Vector2Int you, Vector2Int other)
    {
        return 0f; // placeholder for testing
    }

    public static int[,] FillBoard(int[,] board, Vector2Int p1, Vector2Int p2)
    {
        int width = board.GetLength(0);
        int height = board.GetLength(1);

        int[,] ownership = new int[width, height];
        int[,] dist1 = new int[width, height];
        int[,] dist2 = new int[width, height];

        const int INF = int.MaxValue;

        // Initialize distances
        for (int x = 0; x < width; x++)
        {
            for (int y = 0; y < height; y++)
            {
                dist1[x, y] = INF;
                dist2[x, y] = INF;
            }
        }

        // BFS for Player 1
        FloodFill(board, p1, dist1);

        // BFS for Player 2
        FloodFill(board, p2, dist2);

        int region1 = 0;
        int region2 = 0;

        for (int x = 0; x < width; x++)
        {
            for (int y = 0; y < height; y++)
            {
                if (board[x, y] != 0)
                    continue;

                int d1 = dist1[x, y];
                int d2 = dist2[x, y];

                if (d1 == INF && d2 == INF) {
                    // Debug.Log($"Unreachable free cell at {x},{y}");
                    continue;
                }

                if (d1 < d2) {
                    region1++;
                    ownership[x, y] = 1;
                }
                else if (d2 < d1) {
                    region2++;
                    ownership[x, y] = 2;
                }
                else {
                    ownership[x, y] = 3; // battlefront
                }
            }
        }

        return ownership;
    }
    
    public static int ComputeVoronoiScore(int[,] board, Vector2Int p1, Vector2Int p2)
    {
        int width = board.GetLength(0);
        int height = board.GetLength(1);

        int[,] dist1 = new int[width, height];
        int[,] dist2 = new int[width, height];

        const int INF = int.MaxValue;

        // Initialize distances
        for (int x = 0; x < width; x++)
        {
            for (int y = 0; y < height; y++)
            {
                dist1[x, y] = INF;
                dist2[x, y] = INF;
            }
        }

        // BFS for Player 1
        FloodFill(board, p1, dist1);

        // BFS for Player 2
        FloodFill(board, p2, dist2);

        int region1 = 0;
        int region2 = 0;

        for (int x = 0; x < width; x++)
        {
            for (int y = 0; y < height; y++)
            {
                if (board[x, y] != 0)
                    continue;

                int d1 = dist1[x, y];
                int d2 = dist2[x, y];

                if (d1 == INF && d2 == INF)
                    continue;

                if (d1 < d2)
                    region1++;
                else if (d2 < d1)
                    region2++;
                // equal distance = battlefront → ignored
            }
        }

        return region1 - region2;
    }

    static void FloodFill(int[,] board, Vector2Int start, int[,] dist)
    {
        int width = board.GetLength(0);
        int height = board.GetLength(1);

        Queue<Vector2Int> queue = new Queue<Vector2Int>();

        dist[start.x, start.y] = 0;
        queue.Enqueue(start);

        while (queue.Count > 0)
        {
            Vector2Int current = queue.Dequeue();
            foreach (var dir in DIRS)
            {
                Vector2Int next = current + dir;

                if (next.x < 0 || next.x >= width ||
                    next.y < 0 || next.y >= height)
                    continue;

                if (board[next.x, next.y] != 0)
                    continue;

                if (dist[next.x, next.y] != int.MaxValue)
                    continue;

                dist[next.x, next.y] = dist[current.x, current.y] + 1;
                queue.Enqueue(next);
            }
        }
    }
}

File: rl_core/tron_env/tron_env/heuristic.py
from collections import deque
import numpy as np

INF = float("inf")
DIRS = [(0, -1), (1, 0), (0, 1), (-1, 0)]  # up, right, down, left

def get_best_action(state):
    """ Returns the action that maximizes voronoi score. """
    trails, you, other = state

    best_score = -np.inf
    best_action = None
    for action, dir in enumerate(DIRS):
        new_pos = you + dir

        if not(0 <= new_pos[0] < trails.shape[1]) or \
            not(0 <= new_pos[1] < trails.shape[0]) or \
            trails[new_pos[1], new_pos[0]] != 0:
            continue

        score = voronoi(trails, new_pos, other)

        if score > best_score:
            best_score = score
            best_action = action

    if best_action is None:
        return 0  # If no valid moves, just pick up (or any default)
    return best_action

def flood_fill(trails, start, dist):
    height, width = trails.shape
    queue = deque()

    x0, y0 = start
    dist[y0, x0] = 0
    queue.append((y0, x0))

    while queue:    
        y, x = queue.popleft()

        for dx, dy in DIRS:
            nx, ny = x + dx, y + dy

            if nx < 0 or nx >= width or ny < 0 or ny >= height:
                continue

            if trails[ny, nx] != 0:
                continue

            if dist[ny, nx] != INF:
                continue

            dist[ny, nx] = dist[y, x] + 1
            queue.append((ny, nx))

def get_territories(trails, p1, p2):
    height, width = trails.shape

    dist1 = np.full((height, width), INF)
    dist2 = np.full((height, width), INF)
    territories = np.zeros((height, width), dtype=np.int8)

    flood_fill(trails, p1, dist1)
    flood_fill(trails, p2, dist2)

    for y in range(height):
        for x in range(width):

            if trails[y, x] != 0:
                continue

            d1 = dist1[y, x]
            d2 = dist2[y, x]

            if d1 == INF and d2 == INF:
                continue

            if d1 < d2:
                territories[y, x] = 1
            elif d2 < d1:
                territories[y, x] = 2
            else:
                territories[y, x] = 3  # Battlefront
            
    return territories

def voronoi(trails, p1, p2):
    height, width = trails.shape

    dist1 = np.full((height, width), INF)
    dist2 = np.full((height, width), INF)

    flood_fill(trails, p1, dist1)
    flood_fill(trails, p2, dist2)

    region1 = 0
    region2 = 0

    for y in range(height):
        for x in range(width):

            if trails[y, x] != 0:
                continue

            d1 = dist1[y, x]
            d2 = dist2[y, x]

            if d1 == INF and d2 == INF:
                continue

            if d1 < d2:
                region1 += 1
            elif d2 < d1:
                region2 += 1
            # equal distance → ignored

    return region1 - region2